### PR TITLE
BUG: in _nsorted for frame with duplicated values index

### LINKED ISSUE #13412: BUG: in _nsorted for frame with duplicated values index
The function below has been incorrectly implemented. If the frame has an index with duplicated values, you will get a result with more than `n` rows and not properly sorted. So `nsmallest` and `nlargest` for DataFrame doesn't return a correct frame in this particular case.

```
def _nsorted(self, columns, n, method, keep):
    if not com.is_list_like(columns):
        columns = [columns]
    columns = list(columns)
    ser = getattr(self[columns[0]], method)(n, keep=keep)
    ascending = dict(nlargest=False, nsmallest=True)[method]
    return self.loc[ser.index].sort_values(columns, ascending=ascending,
                                           kind='mergesort')
```

### PR DESCRIPTION
**Pull request by Tux1**
- [x] closes #13412
- [ ] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by Tux1** [comment]
what doesn't pass tests ?

**Comment by sinhrks** [comment]
You must fix test to pass ``flake8`` check.

File: doc/source/whatsnew/v0.19.0.txt [hunk 1]
```diff
@@ -81,3 +81,5 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+
+- Bug in ``DataFrame._nsorted`` when data-frame has duplicated value index. (:issue:`13412`)
```
Comment (sinhrks) [inline doc/source/whatsnew/v0.19.0.txt]: Describe the problem from user's point of view. Users don't care ``_nsorted``.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3315,9 +3315,14 @@ def _nsorted(self, columns, n, method, keep):
         if not com.is_list_like(columns):
             columns = [columns]
         columns = list(columns)
-        ser = getattr(self[columns[0]], method)(n, keep=keep)
+        index_names = self.index.names
```
Comment (sinhrks) [inline pandas/core/frame.py]: Use ``index.is_unique`` to split path.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3315,9 +3315,14 @@ def _nsorted(self, columns, n, method, keep):
...
+        nl = self.index.nlevels
+        df = self.reset_index()
+        ser = getattr(df[columns[0]], method)(n, keep=keep)
         ascending = dict(nlargest=False, nsmallest=True)[method]
-        return self.loc[ser.index].sort_values(columns, ascending=ascending,
+        df = df.loc[ser.index].set_index(df.columns[:nl].tolist())
```
Comment (sinhrks) [inline pandas/core/frame.py]: what's this?

File: pandas/tests/frame/test_analytics.py [hunk 4]
```diff
@@ -1253,6 +1253,21 @@ def test_nsmallest_multiple_columns(self):
         expected = df.sort_values(['a', 'c']).head(5)
         tm.assert_frame_equal(result, expected)
 
+    def test_nsorted_duplicated_index(self):
+        # GH13412
+        df = pd.DataFrame({'a':[1,2,3,4], 'b':[4,3,2,1]}, index=[0,0,1,1])
+        result = df.nlargest(1, 'a')
```
Comment (sinhrks) [inline pandas/tests/frame/test_analytics.py]: Pls also check input larger than 1.

File: pandas/core/frame.py (hunk 3 above)
Comment (Tux1) [inline pandas/core/frame.py]: the only way I found to revert the index correctly. It's not very elegant

File: pandas/tests/frame/test_analytics.py (hunk 4 above)
Comment (Tux1) [inline pandas/tests/frame/test_analytics.py]: In the same test ?

**Comment by TomAugspurger** [comment]
`git diff upstream/master | flake8 --diff ` will help you track down the changes you have to make.

**Comment by jreback** [comment]
I think a simple soln will work here

```
(Pdb) p self.loc[ser.index].head(n).sort_values(columns, ascending=ascending,kind='mergesort')
   a  b
1  3  2
(Pdb) p self.loc[ser.index]
   a  b
1  3  2
1  4  1
```

**Comment by Tux1** [comment]
@jreback what do you mean ? in `_nsorted` ? I don't think so

**Comment by jreback** [comment]
maybe the ``.head()`` should go at the end. I don't recall the exact guarantees of this.

**Comment by Tux1** [comment]
I tested and your soln doesn't work and doesn't pass the test with this case :
```        df = pd.DataFrame({'a': [1, 2, 3, 4],
                           'b': [4, 3, 2, 1]}, index=[0, 0, 1, 1])```

Any other suggestion ?

**Comment by jreback** [comment]
@Tux1 well then play around with it. What you are doing is WAY too complicated for a simple take.

**Comment by jreback** [comment]
can you rebase / update?

**Pull request closed without merging**